### PR TITLE
Avoid nested jitted functions

### PR DESCRIPTION
**Pull request by inailuig**
depends on https://github.com/netket/netket/pull/515

see https://github.com/google/jax/issues/284#issuecomment-465491907

### CODE REVIEW

File: netket/machine/jax.py [hunk 1]
```diff
@@ -231,6 +231,10 @@ def log_val(self, x, out=None):
     def jax_forward(self):
         return self._forward_fn
 
+    @property
+    def jax_forward_nj(self):
+        return self._forward_fn_nj
```
Comment (PhilipVinc) [inline netket/machine/jax.py:236]: Do we really need to export this? 
I think we can keep it as a somewhat-private thing used inside of netket.

Comment (inailuig) [reply netket/machine/jax.py:236]: I would actually need it to do the jvp/vjp 's in https://github.com/netket/netket/pull/517
but I agree that its not optimal.
Maybe its better to just just access the private member there?

Comment (PhilipVinc) [reply netket/machine/jax.py:236]: yep.
I'd say let's just not make it part of the public front facing API. 
Keep the field `_forward_fn_nj` but remove the `jax_forward_nj` property.

Comment (PhilipVinc) [inline netket/machine/jax.py:236]:
```suggestion
```

**Review by PhilipVinc — CHANGES_REQUESTED** [review]

**Comment by PhilipVinc** [comment]
For reference (would be nice to do it at some point):
I believe we could move the `forward_scalar` closure outside of the `Jax` class constructor together with all the logic that builds the jitted vip/jvp/perex functions.

That way, if one uses the same jax module but creates a new jax machine, all those methods are already jitted and won't be rejitted

File: netket/sampler/jax/metropolis_hastings.py [hunk 2]
```diff
@@ -132,7 +132,7 @@ def generate_samples(self, n_samples, init_random=False, samples=None):
         self.reset(init_random)
 
         self._rng_key, samples = self._metropolis_kernel(
-            self.machine.jax_forward_nj,
+            self.machine._forward_fn_nj,
```
Comment (inailuig) [inline netket/sampler/jax/metropolis_hastings.py:135]: @PhilipVinc so from now on we just access the non-jitted forward function like this whenever we need it internally?

Comment (PhilipVinc) [reply netket/sampler/jax/metropolis_hastings.py:135]: yes.

**Comment by gcarleo** [comment]
Is this impacting performance in any significant way (for better or worse) ?

**Review by PhilipVinc — APPROVED** [review]

**Pull request merged**